### PR TITLE
refactor WebsocketFrame and support receiving frames in multiple chunks

### PR DESCRIPTION
**Pull request by jodevsa**
Hi @KhafraDev,

I have worked on the following: 

* Add support to build the frame from multiple chunks
* Refactor and improve the WebsocketFrame class
* fix some bugs
* add a method to calculate the expected size of the frame when it's converted to a buffer ( this will be useful later to be able to detect if the chunk contains multiple frames)

### CODE REVIEW

File: lib/websocket/connection.js [hunk 1]
```diff
@@ -280,71 +285,95 @@ class WebsocketFrame {
...
 
     for (let i = 0; i < 4; i++) {
       this.maskKey[i] = Math.floor(Math.random() * 256)
     }
   }
 
-  static from (buffer) {
-    let nextByte = 0
-    const frame = new WebsocketFrame()
+  size () {
```
Comment (jodevsa) [inline lib/websocket/connection.js]: This is the size when the frame is converted to a buffer. I don't like the name so any suggestions would be great

File: lib/websocket/connection.js [hunk 2]
```diff
@@ -246,32 +245,38 @@ class WebsocketFrame {
...
+      this.mask()
+    }
   }
 
   toBuffer () {
-    // TODO: revisit the buffer size calculaton
-    const payloadLength = this.payloadData?.length ?? 0
-    const buffer = Buffer.alloc(4 + 2 + payloadLength)
-
+    const buffer = Buffer.alloc(this.size())
```
Comment (KhafraDev) [inline lib/websocket/connection.js]: allocUnsafe? If we're overwriting every byte in the buffer.

File: lib/websocket/connection.js (hunk 1 above)
Comment (KhafraDev) [reply lib/websocket/connection.js]: maybe something like byteLength to match existing apis?

File: lib/websocket/connection.js [hunk 3]
```diff
@@ -357,9 +386,18 @@ class WebsocketFrame {
  */
 function receiveData (ws) {
   const { [kResponse]: response } = ws
-
+  let buffer = Buffer.alloc(0)
   response.socket.on('data', (chunk) => {
-    const frame = WebsocketFrame.from(chunk)
+    buffer = Buffer.concat([buffer, chunk])
```
Comment (KhafraDev) [inline lib/websocket/connection.js:393]: If we can get the payload length from the first chunk sent to us, it should be possible to allocate the buffer of that size once we receive the length, and then use `buffer.set` to "concatenate" them as we receive.

**Review by KhafraDev — CHANGES_REQUESTED** [review]

File: lib/websocket/connection.js (hunk 3 above)
Comment (jodevsa) [reply lib/websocket/connection.js:393]: makes sense. can we have a case where we get a chunk with the payload length still not in?

Comment (jodevsa) [reply lib/websocket/connection.js:393]: I've added a TODO and I'll work on it soon, do you think we can merge this for now?

Comment (jodevsa) [reply lib/websocket/connection.js:393]: btw, do you think it's a good idea to start working on some tests or is it still too early?

Comment (KhafraDev) [reply lib/websocket/connection.js:393]: If any of the tests pass from https://github.com/web-platform-tests/wpt/tree/master/websockets, we can add them. We don't necessarily need to add our own, but it won't harm anyone if you did. 👍

Comment (KhafraDev) [reply lib/websocket/connection.js:393]: Actually I doubt any of those tests will work until we can setup a WebSocket server.

**Comment by KhafraDev** [comment]
@jodevsa Hey, would you mind checking something out for me? We have some issues with reading the status code and reason for the close frames randomly, which causes the below test to fail (maybe every 10th run).

https://raw.githubusercontent.com/web-platform-tests/wpt/master/websockets/Close-1005-verify-code.any.js

All you need to run it is to download that file and put it in the `test/wpt/tests/websockets` folder. Afterwards you can run it by doing `node test/wpt/start-websockets.mjs`. If you're on windows, you can run the tests multiple times in powershell by doing `1..100 | % { node test/wpt/start-websockets.mjs }` (theoretically all 100 runs should pass).

I've been banging my head at this for hours trying to fix it without luck. The status code (& reason) are being sent to the server correctly, but either the server is echoing back the wrong data, or we are parsing it incorrectly.

Thanks!

---

I will also likely rewrite the WebsocketFrame class at some point and split it into 2 different classes (one for sending data, and one for receiving). Should make working with it easier, and will make parsing fragmented frames easier.

**Comment by KhafraDev** [comment]
Never mind, I found a fix 🤦.

```diff
diff --git a/test/wpt/server/websocket.mjs b/test/wpt/server/websocket.mjs
index a90ed89..3d419cd 100644
--- a/test/wpt/server/websocket.mjs
+++ b/test/wpt/server/websocket.mjs
@@ -12,6 +12,4 @@ wss.on('connection', (ws) => {
   ws.on('message', (data) => {
     ws.send(data)
   })
-
-  ws.send('Connected!')
 })
```

This still seems like an issue in the implementation, but for now this works. Maybe the socket receives data after being set to a `closing` state?

**Pull request merged**